common/flatpak-transaction-log.c

```diff
@@ -0,0 +1,305 @@
+
+  json_builder_begin_array (builder);
+  json_builder_end_array (builder);
+
+  return json_builder_get_root (builder);
+}
+
+static JsonNode *
+load_existing_log_if_exists (GFile        *path,
+                             GCancellable *cancellable,
+                             GError       **error)
+{
+  JsonNode *toplevel = NULL;
+  g_autoptr(JsonNode) array = NULL;
+  g_autoptr(GError) my_error = NULL;
+
+  toplevel = load_existing_log_from_file (path, cancellable, &my_error);
+
+  if (!toplevel)
+    {
+      if (!g_error_matches (my_error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+        {
+          g_propagate_error (error, g_steal_pointer (&my_error));
+          return NULL;
+        }
+
+      toplevel = build_empty_array ();
+    }
+  
+  return toplevel;
+}
+
+static gboolean
+write_node_to_log_file (GFile        *file,
+                        JsonNode     *node,
+                        GCancellable *cancellable,
+                        GError       **error)
+{
+  g_autoptr(GFile) parent = g_file_get_parent (file);
+  g_autoptr(GFileOutputStream) stream = NULL;
+  g_autoptr(JsonGenerator) generator = NULL;
+  g_autoptr(GError) my_error = NULL;
+
+  if (!g_file_make_directory_with_parents (parent, cancellable, &my_error))
+    {
+      if (!g_error_matches (my_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        {
+          g_propagate_error (error, g_steal_pointer (&my_error));
+          return FALSE;
+        }
+    }
+
+  stream = g_file_replace (file,
+                           NULL,
+                           TRUE,
+                           G_FILE_CREATE_NONE,
+                           cancellable,
+                           error);
+
+  if (!stream)
+    return FALSE;
+
+  generator = json_generator_new ();
+  json_generator_set_root (generator, node);
+  json_generator_set_pretty (generator, TRUE);
+
+  if (!json_generator_to_stream (generator, G_OUTPUT_STREAM (stream), cancellable, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+#define EVENT_LOG_VERSION "1"
+
+static JsonNode *
+build_base_event (const gchar *event)
+{
+  g_autoptr(JsonBuilder) builder = json_builder_new ();
+
+  json_builder_begin_object (builder);
+
+  json_builder_set_member_name (builder, "event");
+  json_builder_add_string_value (builder, event);
+
+  json_builder_set_member_name (builder, "version");
+  json_builder_add_string_value (builder, EVENT_LOG_VERSION);
+  json_builder_end_object (builder);
+
+  return json_builder_get_root (builder);
+}
+
+static JsonNode *
+deploy_install_event (const gchar *ref,
+                      const gchar *origin)
+{
+  JsonNode *base = build_base_event ("deploy-install");
+  JsonObject *object = json_node_get_object (base);
+
+  json_object_set_string_member (object, "ref", ref);
+  json_object_set_string_member (object, "origin", origin);
+
+  return base;
+}
+
+gboolean
+flatpak_transaction_log_commit_deploy_install_event (FlatpakTransactionLog *log,
```

Comment:
We only log deployments of installations as opposed to updates or appstream updates. Not sure if we want to log those under the some event type or log them separately (or not at all).

Reply:
You're the primary consumer of this, but i think its probably useful to record every deploy, rather than just the initial install.

Reply:
Maybe with a "previous deployed version" property, which is not there if this is the initial install.